File: backend/apps/executions/api/serializers/step.py
from rest_framework import serializers
from apps.executions.models import Step, Execution
from django.utils import timezone
from apps.executions.constants import ExecutionStatus


class ExecutionStepSerializer(serializers.ModelSerializer):
    class Meta:
        model = Execution
        fields = [
            "status",
            "date_start",
            "date_end"
        ]


class StepSerializer(serializers.ModelSerializer):
    execution_data = ExecutionStepSerializer(
        read_only=True, source='execution')

    class Meta:
        model = Step
        fields = [
            'id',
            "identification",
            "status",
            "date_start",
            "date_end",
            "time_automation_seconds",
            "time_manual_seconds",
            "time_economy_seconds",
            "potential_time_seconds",
            "potential_cost",
            "efficiency_percent",
            "execution",
            "execution_data"

        ]
        read_only_fields = [
            "date_start",
            "date_end",
            "time_automation_seconds",
            "time_manual_seconds",
            "time_economy_seconds",
            "potential_time_seconds",
            "potential_cost",
            "efficiency_percent",
            "execution_data"
        ]

    def update(self, instance, validated_data):
        status = validated_data.get('status', instance.status)

        if status in [ExecutionStatus.CONCLUIDO, ExecutionStatus.ALERTA] and instance.date_end is None:
            instance.date_end = timezone.now()

        instance.status = status
        return super().update(instance, validated_data)

File: frontend/src/types/step.ts
export type StepStatus = 'iniciado' | 'concluido' | 'erro' | 'alerta' | 'teste';

export interface Step {
  id: number;
  identification: string | null;
  status: StepStatus;
  date_start: string;
  date_end: string | null;
  time_automation_seconds: number;
  time_manual_seconds: number;
  time_economy_seconds: number;
  potential_time_seconds: number;
  potential_cost: string;
  efficiency_percent: string;
  execution: number;
  execution_data: { status: string; date_start: string; date_end: string | null };
}

export interface StepCreate {
  execution: number;
  identification?: string;
  status: StepStatus;
}

export interface StepFilters {
  execution?: string;
  identification?: string;
  status?: string;
  date_start?: string;
}
